src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetRemoteFileService.cs

```diff
@@ -0,0 +1,264 @@
+        private IServiceBroker _serviceBroker;
+        private AuthorizationServiceClient? _authorizationServiceClient;
+        private INuGetTelemetryProvider _nuGetTelemetryProvider;
+        private bool _disposedValue;
+
+        internal readonly static MemoryCache IdentityToUriCache = new MemoryCache("PackageSearchMetadata",
+            new NameValueCollection
+            {
+                { "cacheMemoryLimitMegabytes", "4" },
+                { "physicalMemoryLimitPercentage", "0" },
+                { "pollingInterval", "00:02:00" }
+            });
+
+        private static readonly CacheItemPolicy CacheItemPolicy = new CacheItemPolicy
+        {
+            SlidingExpiration = ObjectCache.NoSlidingExpiration,
+            AbsoluteExpiration = ObjectCache.InfiniteAbsoluteExpiration,
+        };
+
+        public static void AddIconToCache(PackageIdentity packageIdentity, Uri iconUri)
+        {
+            string key = "icon:" + packageIdentity.ToString();
+            if (iconUri != null && !IdentityToUriCache.Add(key, iconUri, CacheItemPolicy))
+            {
+                IdentityToUriCache.Remove(key);
+                IdentityToUriCache.Add(key, iconUri, CacheItemPolicy);
+            }
+        }
+
+        public static void AddLicenseToCache(PackageIdentity packageIdentity, Uri embeddedLicenseUri)
+        {
+            Assumes.NotNull(embeddedLicenseUri);
+            string key = "license:" + packageIdentity.ToString();
+            if (!IdentityToUriCache.Add(key, embeddedLicenseUri, CacheItemPolicy))
+            {
+                IdentityToUriCache.Remove(key);
+                IdentityToUriCache.Add(key, embeddedLicenseUri, CacheItemPolicy);
+            }
+        }
+
+        public NuGetRemoteFileService(
+            ServiceActivationOptions options,
+            IServiceBroker serviceBroker,
+            AuthorizationServiceClient authorizationServiceClient,
+            INuGetTelemetryProvider nuGetTelemetryProvider)
+        {
+            _options = options;
+            _serviceBroker = serviceBroker;
+            _authorizationServiceClient = authorizationServiceClient;
+            _nuGetTelemetryProvider = nuGetTelemetryProvider;
+            Assumes.NotNull(_serviceBroker);
+            Assumes.NotNull(_authorizationServiceClient);
+            Assumes.NotNull(_nuGetTelemetryProvider);
+        }
+
+        public NuGetRemoteFileService(IServiceBroker serviceBroker, INuGetTelemetryProvider nuGetTelemetryProvider)
+        {
+            _serviceBroker = serviceBroker;
+            Assumes.NotNull(_serviceBroker);
+
+            _nuGetTelemetryProvider = nuGetTelemetryProvider;
+            Assumes.NotNull(_nuGetTelemetryProvider);
+        }
+
+        public async ValueTask<Stream?> GetPackageIconAsync(PackageIdentity packageIdentity, CancellationToken cancellationToken)
+        {
+            Assumes.NotNull(packageIdentity);
+            string key = "icon:" + packageIdentity.ToString();
```

Comment:
I think it's better to use string constants instead of hardcoded strings:

```suggestion
            string key = NuGetRemoteFileService.IconPrefix + packageIdentity.ToString();
```

Reply:
fixed in next iteration.